core/opendaq/reader/src/block_reader_impl.cpp

```diff
@@ -163,27 +196,56 @@ ErrCode BlockReaderImpl::readPacketData()
         }
     }
 
-    if (toRead < remainingSampleCount)
+    info.writtenSampleCount += sampleCountToRead;
+    info.prevSampleIndex += sampleCountToRead;
+    info.remainingSamplesToRead -= sampleCountToRead;
+
+    if (info.writtenSampleCount % blockSize == 0)
     {
-        info.prevSampleIndex += toRead;
+        // how many data packet need to be stepped back
+        const auto rewindSamples = overlappedBlockSize;
+        auto rewindPackets = rewindSamples / packetSampleCount;
+        auto rewindPacketRemainder = rewindSamples % packetSampleCount;
+
+        if (info.prevSampleIndex < rewindPacketRemainder)
+        {
+            rewindPackets += 1;
+            info.prevSampleIndex = packetSampleCount - rewindPacketRemainder;
+        }
+        else
+        {
+            info.prevSampleIndex -= rewindPacketRemainder;
+        }
+
+        using IterDiff = BlockReadInfo::DataPacketsQueueType::difference_type;
+
+        info.currentDataPacketIter = std::next(info.currentDataPacketIter, -static_cast<IterDiff>(rewindPackets));
+
+        if (info.remainingSamplesToRead)
+            info.remainingSamplesToRead += overlappedBlockSize;
     }
-    else
-    {
-        std::unique_lock lock(notify.mutex);
-        notify.dataReady = false;
 
+    if (info.prevSampleIndex == packetSampleCount)
+    {
+        if (++info.currentDataPacketIter == info.dataPacketsQueue.end())
+        {
+            std::unique_lock lock(notify.mutex);
+            notify.dataReady = false;
+        }
         info.reset();
```

Comment:
Reset is maybe not the most apt naming for the method, as it does not actually reset the info - the packet queue is still kept.

Reply:
Renamed